### ISSUE TITLE
Performous crash while starting a duet song, if the microphone devices are not connected

### ISSUE THREAD

**Issue by Stefan1200-de** [performous/performous#255]
### Actual behaviour

If you try to start a duet game while you forgot to connect the USB microphones, Performous will crash to desktop.

### Expected behaviour

Giving an error message, that the microphone devices are missing and allow starting the song without microphones, to test the txt file.

### Details

Provide some additional information:

- performous version: 1.1
- operating system + version: Windows 10 x64

**Comment by Baklap4** [comment]
I can't test / wouldn't know how to test this on my linux laptop, since i have no clue on how to disable my soundcard's internal mic.. @nieknooijens ?

**Comment by nieknooijens** [comment]
I also think this is pretty hard to debug since it's probably something related to WASAPI

performous binds to portaudio which binds to whatever the heck is available on that platform. in the case of windows that's directSound/WASAPI and in the case of linux that's ALSA/Pulseaudio/Jack.

since Jack also works on windows you could try using that sound API (run JACKD) and see if it still crashes when you add devices while performous is running (just to check whether it's performous which can't deal with variable amounts of devices, or the sound API glitching out). 
see jackaudio.org

**Comment by Stefan1200-de** [comment]
Just in case that this debug log helps:
[infolog.txt](https://github.com/performous/performous/files/1107846/infolog.txt)

But I found nothing crash related. As soon as I connect both USB microphones, Performous does not crash anymore.

**Comment by MagicChocoRolls** [comment]
I had the very same problem on my Mac and can reproduce it the following way:

* Configure at least 2 microphones (aka players) on a detachable sound card
* start a song and verify it is working
* close performous, disconnect the microphone you configured (this only works for detachable equipment like the USB sing star mics)
* restart performous and go straight to song selection. A duet song will crash the application

This happens because the sound card has disappeared and the players are set back to defaults. You have to reconfigure them and it will work fine again.
Maybe this is enough to fix the problem. I assume it is enough to check if enough players are configured for a duet song ;-)

Regards
Bondfreak

**Comment by OznOg** [comment]
I tried this on Linux on the current master (1d96a6aebcb44d62803e5c36626a7580b15e862f) and does not seem to crash. Thus maybe it is not there anymore, or maybe it is a windows/mac related issue...
I know this was a long time ago but could you tell me if you still experience this bug on recent versions?

**Comment by Stefan1200-de** [comment]
> I know this was a long time ago but could you tell me if you still experience this bug on recent versions?

Well, Performous 1.1 is still the recent release ;)
Currently I don't use Performous anymore (I use another Software, which is compatible with that songs). Maybe I do another test, if there is a new Windows release.

**Comment by Baklap4** [comment]
Not quite sure but i do get it to crash on windows, but not persé with or without microphones.

The message is "invalid comparator" and it happens upon the `std::merge()` in `songparser-txt.cc`:

```
std::merge(s1.begin(), s1.end(), s2.begin(), s2.end(), std::back_inserter(merged), Note::ltBegin);
```
The output
```
Debug Assertion Failed!

Program: ...e\repos\performous\performous2\build\x64-Debug\Performous.exe
File: C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30037\include\xutility
Line: 1520

Expression: invalid comparator

For information on how your program can cause an assertion
failure, see the Visual C++ documentation on asserts.

(Press Retry to debug the application)
'Performous.exe' (Win32): Loaded 'C:\Windows\System32\TextShaping.dll'. 
Performous.exe has triggered a breakpoint.

The program '[14844] Performous.exe' has exited with code 0 (0x0)
```

**Comment by Baklap4** [comment]
Bug is fixed in #650

**Comment by Baklap4** [comment]
@Stefan1200-de This is fixed within latest master. A new release will soon come. Join the conversation here: https://github.com/performous/performous/discussions/665
